solvation_analysis/analysis.py

```diff
@@ -0,0 +1,127 @@
+        """
+        Parameters
+        ----------
+        solute (AtomGroup): an atom group  # TODO: force to atom group
+        solvents (dict): a dictionary of names and atom groups.
+            e.g. {"name_1": solvent_group_1, "name_2": solvent_group_2, ...}
+        radii (dict): an optional dictionary of solvation radii, any radii not
+            given will be calculated. e.g. {"name_2": radius_2, "name_5": radius_5}
+        rdf_kernel (func): this function must take rdf bins and data as input and return
+            a solvation radius as output. e.g. rdf_kernel(bins, data) -> 3.2. By default,
+            the rdf_kernel is solvation_analysis.rdf_parser.identify_solvation_cutoff.
+        kernel_kwargs: kwargs passed to rdf_kernel
+        kwargs: kwargs passed to AnalysisBase
+        """
+        super(Solute, self).__init__(solute.universe.trajectory, **kwargs)
+        self.radii = {} if radii is None else radii
+        self.kernel = identify_solvation_cutoff if rdf_kernel is None else rdf_kernel
+        # if not kernel:
+        #     self.kernel = identify_solvation_cutoff
+        self.kernel_kwargs = {} if kernel_kwargs is None else radii
+        # if not kernel_kwargs:
+        #     self.kernel_kwargs = {}
+
+        self.solute = solute
+        self.solvents = solvents
+        self.u = self.solute.universe
+        self.rdf_plots = {}
+        self.rdf_data = {}
+        self.ion_speciation = None
+        self.ion_pairing = None
+        self.coordination_numbers = None
+        self.solvation_frames = []
+
+    @classmethod
+    def _plot_solvation_radius(cls, bins, data, radius):
+        fig, ax = plt.subplots()
+        ax.plot(bins, data, "b-", label="rdf")
+        ax.axvline(radius, color="r", label="solvation radius")
+        ax.set_xlabel("Radial Distance (A)")
+        ax.set_ylabel("Probability Density")
+        ax.legend()
+        return fig, ax
+
+    def run_prepare(self):
+        for name, solvent in self.solvents.items():
+            # generate and save RDFs
+            rdf = InterRDF(self.solute, solvent, range=(0.0, 8.0))
+            # TODO: specify start and stop args
+            rdf.run()
+            bins, data = rdf.results.bins, rdf.results.rdf
+            self.rdf_data[name] = (data, bins)
+            # generate and save plots
+            if name not in self.radii.keys():
+                self.radii[name] = self.kernel(bins, data, **self.kernel_kwargs)
+            fig, ax = self._plot_solvation_radius(bins, data, self.radii[name])
+            ax.set_title(f"Solvation distance of {name}")
+            self.rdf_plots[name] = fig, ax
+
+    def _prepare(self):
+        assert self.solvents.keys() == self.radii.keys(), "Radii missing."
+        # columns: solute #, atom id, distance, solvent name, res id
+
+    def _single_frame(self):
+        # initialize empty arrays
+        all_pairs = np.empty((0, 2), dtype=np.int)
+        all_dist = np.empty(0)
+        all_tags = np.empty(0)
```

Comment:
I'm not super versed about this, but is concatenate doing lots of memory re-allocation in the background to ensure contiguous memory? i.e. would it be faster to initially just append to lists and then do a np.array(list) call afterwards?

Reply:
+1 this is where lists will be much faster, I think. Every `np.concatenate` call will copy the previous array as the new array needs to be contiguous. Lists pre-allocate memory chunks at a time.

Can you just do the one `capped_distances` call with all the solutes and then label them afterwards?

Reply:
Other option is to figure out the size of the resulting array, allocate in one shot and then index, but can be a little difficult to follow. The way you have done it is nice and expressive  but +1 on use of lists.

Reply:
This makes sense to me. Resolved in the newest commits.